### PR TITLE
Make client.ContainerWait easier and safer to use

### PR DESCRIPTION
**Pull request by tiborvass**
The caller of this function needed to be extra careful
by making sure they are draining all the return channels
in a specific order to avoid leaking goroutines.

This commit introduces ackChan as a parameter to allow
for the same behavior the previous signature was designed for.

Signed-off-by: Tibor Vass <tibor@docker.com>

### CODE REVIEW

File: client/container_wait.go
Comment (jlhawn) [inline client/container_wait.go]: This is a great example 👍

File: client/container_wait.go
Comment (tonistiigi) [inline client/container_wait.go]: Shouldn't this be called only if there isn't an error

**Review by tonistiigi — APPROVED** [review]
LGTM

**Comment by tonistiigi** [comment]
@tiborvass  unit tests are red in ci

File: client/container_wait.go
Comment (stevvooe) [inline client/container_wait.go]: I think you can fix this problem with by replacing this with:

```
select {
case resultC <-res:
default:
}
```

Or, you can make it buffered.

**Review by jlhawn — APPROVED** [review]

**Comment by tiborvass** [comment]
@tonistiigi @jlhawn do you think like @stevvooe that the old API was fine and I should not change it, because it is a common API for Wait-like functions? If so, i could fix the leak by just buffering the channel like he suggested, and not have any API change.

File: client/container_wait.go
Comment (tiborvass) [reply client/container_wait.go]: I debated the same thing, I believe it's how it was before that's why I chose this. I can change it.

**Comment by stevvooe** [comment]
I think if you're going to this far, you can completely get rid of the channel argument. I can just call this function like this:

```
go func() {
  defer close(ack)
  r, err := ContainerWait(ctx, containerID, condition)
  // ...  
}
```

This is better because it lets the caller decide how much channel stuff is happening. It also ensures that the result and ack are synchronized, avoiding possible races there.

However, I am not sure this change is a great one to make as the refactoring required for users of this function might be fairly involved.

File: client/container_wait.go
Comment (tonistiigi) [reply client/container_wait.go]: In the old code it seems to be after error check.

**Comment by jlhawn** [comment]
If buffering the channel would not be considered "bad design" or "code smell" like I have been told in the past, then I'd prefer that.

**Comment by jlhawn** [comment]
> I think if you're going to this far, you can completely get rid of the channel argument. I can just call this function like this:

     go func() {
          defer close(ack)
          r, err := ContainerWait(ctx, containerID, condition)
          // ...  
    }

> This is better because it lets the caller decide how much channel stuff is happening. It also ensures that the result and ack are synchronized, avoiding possible races there.

I think you're misunderstanding what the ack we're talking about is for. It's not for knowing when the Wait call unblocks (that was always possible with the example you show). The ack is to know when the API server has *started* waiting.

**Comment by stevvooe** [comment]
> If buffering the channel would not be considered "bad design" or "code smell" like I have been told in the past, then I'd prefer that.

Buffering a channel is not a code smell and I don't know where this came from. It is useful as a "dropbox" for a result that decouples synchronization of sender and receiver.

> The ack is to know when the API server has started waiting.

This might the source of the problem. I'm also not clear on the value of the distinction. In general, a Wait function should be a blocking Wait and should only return when the condition for which are waiting is satisfied or when there is an error or timeout. The wait call should be considered to start at the function call boundary.

**Comment by tonistiigi** [comment]
Buffering a channel is not a problem (unless it is used as a bypass to avoid deadlocks because then it in many cases just reduces scope instead of fixing a problem but that is not an issue here, yet). Returning channels is another topic. I consider it a bad practice as it makes unclear how to control the lifetime of the call. It is unclear who is responsible for that channel, if these channels need to be read(issue here), do you need to wait for these channels to close for cleanup, is the context passed to this function reusable in other calls(no in this case), how to do safe cleanup from client side(in this case, cancel the context, then wait one of the channels to return). In case of multiple returns, do both of them need to be read, in what order? There is no way to know how the previous API method should be used without carefully reading every line of its implementation. If returning channels is actually required I'd suggest something like stdlib timer that returns a wrapper around channel and lifetime can be controlled with the methods of this wrapper.

File: client/container_wait.go
Comment (tonistiigi) [reply client/container_wait.go]: We should not skip sending the result if a receiver is not ready yet.

**Comment by jlhawn** [comment]
> > The ack is to know when the API server has *started* waiting.
> 
> This might the source of the problem. I'm also not clear on the value of the distinction.

`docker run ...` will make this series of API requests:

- `ContainerCreate`
- `ContainerAttach`
- `ContainerWait` (for it to exit)
- `ContainerStart`

Then it blocks on the `ContainerWait()` result channel (if not in detached mode).

If a `ContainerWait()` call blocked until the wait condition was met then it would have to be done in a separate goroutine that sends the result on a channel. Past versions of Docker did this. The problem with that is that it is followed immediately by a `ContainerStart` call in the main goroutine.

Now there's a race condition between the `Wait` and `Start` API calls: if the container starts and exits *before* the API server handles the `Wait` call with a `next-exit` wait condition, then the `Wait` call will not return until the container is either started again and exits or is removed.

In v1.30, we have the `Wait` API handler return response headers immediately after it has a channel it will use to block on the wait result. It then returns the rest of the response body with the result once it has it. 

The current API client method blocks only until it gets the response header back, then returns a channel from which you can receive the wait result once the response body is returned after the container actually exits. At this point, we make the `ContainerStart` API call and then block on receiving from the channel. This avoids the race between `ContainerWait` and `ContainerStart`.

**Comment by stevvooe** [comment]
@jlhwan It seems like we are missing a piece to identify the waiting context. For example, let's say we have wait api like this:

```
p := Start()
Wait(p)
```

Such an API doesn't race because we can uniquely identify the call to start with some return value `p`. If start is called again, between `Start` and `Wait`, `p` gets a different value (let's say `q`) and the race is avoided.

More concretely, do we have something that we can use to identify the source of the start call to avoid the race?

**Comment by vdemeester** [comment]
ping @tiborvass what is the status here 👼

**Comment by thaJeztah** [comment]
@tiborvass - discussing in the maintainers meeting we're closing this for now, but let us know if you want to work on this again

**Pull request closed without merging**
